### PR TITLE
Project todos - Anna Hellqvist

### PR DESCRIPTION
**Pull request by AnnaHellqvist**
Link to my deployed project:
https://annas-todo-list.netlify.app/

My reflections on how this week's project turned out:
I thought this was a fun project and week. I felt like I got a better understanding of redux.
If I had more time I would have put some more effort into the styling.. I did'nt have the energy or any inspiration. Will probably change this is the future. I would also have wanted to try to add a local storage.

### CODE REVIEW

File: code/src/App.js [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
   return (
-    <div>
-      Find me in src/app.js!
-    </div>
+    <Provider store={store}>
+      <Header />
+      <AddTask />
+      <TaskList />
+      <Footer />
+    </Provider>
```
Comment (petrame) [inline code/src/App.js:22]: Nice and clean with the division into components.

File: code/src/components/AddTask.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    dispatch(tasks.actions.addTask({ id, text, isComplete }));
+    setText('');
+  }
+
+  // setText to the current value in textbox
+  //checks length of array of items and sets id to the next in line, if length is 0 set id to 1
+  const onChangeEvent = (value) => {
+    setText(value);
+    setId(allTasksArray.length === 0 ? 1 : allTasksArray[allTasksArray.length-1].id + 1);
+  }
```
Comment (petrame) [inline code/src/components/AddTask.js:27]: I found some problems with ID:s when creating empty tasks. The empty task got the same ID as an existing task and when I removed it from list I ended up with a task that couldn't be deleted.

And maybe the id allocation could take place in the reducer?

File: code/src/components/Header.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+const Header = () => {
+  const tasksArray = useSelector((store) => store.tasks.items);
+  const completedTasks = tasksArray.filter((task) => task.isComplete);
+
+  return (
+    <HeaderContainer>
+      <HeaderWrapper>
+        <h1>My todo list</h1>
+        <p>You have completed: {completedTasks.length} / {tasksArray.length} </p>
```
Comment (petrame) [inline code/src/components/Header.js:14]: Nice function ⭐

File: code/src/App.js [hunk 4]
```diff
@@ -1,9 +1,24 @@
 import React from 'react'
+import { Provider } from 'react-redux'
+import { configureStore, combineReducers } from '@reduxjs/toolkit'
+
+import { tasks } from 'reducers/tasks'
+import Header from 'components/Header'
+import TaskList from 'components/TaskList'
+import AddTask from 'components/AddTask'
+import Footer from 'components/Footer'
```
Comment (petrame) [inline code/src/App.js]: Nice separation of the imports.

File: code/src/components/TaskList.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <TaskContainer>
+      {allTasksArray.map(task => {
+        return (
+        <Task key={task.id} task={task} />
+        )
+      })}
+      {allTasksArray.length === 0 && (
+        <PlaceholderWrapper>
+          <img src={Rocket} alt='rocket placeholder'></img>
+          <p>Huurraaay, no more tasks! You are done!</p>
```
Comment (petrame) [inline code/src/components/TaskList.js:18]: Awesome! 🚀

File: code/src/reducers/tasks.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      const filteredItems = state.items.filter(item => item.id !== action.payload);
+      state.items = filteredItems;
+    },
+    toggleCompleteStatus: (state, action) => {
+      // Finds the task/item with same id as payload (the task that was clicked)
+      const choosenTask = state.items.find(item => item.id === action.payload)
+
+      // Toggles between true and false
+      if (choosenTask) {
+        choosenTask.isComplete = !choosenTask.isComplete;
```
Comment (petrame) [inline code/src/reducers/tasks.js:31]: I learned this from Vans code session, I think it's so smart ⭐

File: code/src/styling/styling.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
+import styled from 'styled-components';
```
Comment (petrame) [inline code/src/styling/styling.js:1]: Nice with styled components!

**Review by petrame — APPROVED** [review]
Great job Anna! You managed to do some of the red level requirements 🎉
Your code is clean with good and consistent naming which makes it easy to read.
The spaceship that shows up when the list is empty is a really nice feature, I should have done something similar.

File: code/src/App.js (hunk 4 above)
Comment (EmOlsz) [reply code/src/App.js]: Next time let's remember about `./` in path to file though

File: code/src/components/AddTask.js (hunk 2 above)
Comment (EmOlsz) [reply code/src/components/AddTask.js:27]: Yep, this needs to be fixed - we shouldn't be able to add empty tasks

File: code/src/components/TaskList.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+const TaskList = () => {
+  // fetch the tasks from the store
+  const allTasksArray = useSelector((store) => store.tasks.items)
+
+  return (
+    <TaskContainer>
+      {allTasksArray.map(task => {
+        return (
+        <Task key={task.id} task={task} />
```
Comment (EmOlsz) [inline code/src/components/TaskList.js]: or just `{allTasksArray.map(task => <Task key={task.id} task={task} />}`

File: code/src/components/AddTask.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    event.preventDefault();
+    dispatch(tasks.actions.addTask({ id, text, isComplete }));
+    setText('');
+  }
+
+  // setText to the current value in textbox
+  //checks length of array of items and sets id to the next in line, if length is 0 set id to 1
+  const onChangeEvent = (value) => {
+    setText(value);
+    setId(allTasksArray.length === 0 ? 1 : allTasksArray[allTasksArray.length-1].id + 1);
```
Comment (EmOlsz) [inline code/src/components/AddTask.js:26]: clever!

File: code/src/components/Task.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  const handleCheckboxChange = ({task}) => {
+    dispatch(tasks.actions.toggleCompleteStatus(task.id));
+  }
+  
+  return (
+      <TaskWrapper>
+        <Checkbox type='checkbox' onChange={() => handleCheckboxChange({ task })}></Checkbox>
+        <TaskText>{task.text}</TaskText>
```
Comment (EmOlsz) [inline code/src/components/Task.js]: That could be a label - right now it's not possible to mark task complete without clicking on checkbox only

File: code/src/components/AddTask.js (hunk 2 above)
Comment (AnnaHellqvist) [reply code/src/components/AddTask.js:27]: added required for the text field 👍

File: code/src/App.js (hunk 4 above)
Comment (AnnaHellqvist) [reply code/src/App.js]: fixed!

**Pull request closed without merging**